### ISSUE TITLE
Bug: index out of range [-1] when starting caddy

### ISSUE THREAD

**Issue by psrcek** [caddyserver/caddy#4101]
Upon starting caddy the following is printed in a loop (I'm assuming because my container automatically restarts)
This only happens one one of my 2 servers, with the same config, which is very weird.

```
{"level":"info","ts":1617283700.1680074,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":"caddyfile"}
panic: runtime error: index out of range [-1]

goroutine 1 [running]:
github.com/caddyserver/caddy/v2/caddyconfig/httpcaddyfile.consolidateAutomationPolicies(0xc00052e800, 0x3, 0x40, 0x0, 0x0, 0xc0006e0de0)
	github.com/caddyserver/caddy/v2@v2.4.0-beta.1/caddyconfig/httpcaddyfile/tlsapp.go:498 +0xc4a
github.com/caddyserver/caddy/v2/caddyconfig/httpcaddyfile.ServerType.buildTLSApp(0xc0001f25a0, 0x5, 0x5, 0xc0006bb800, 0x0, 0x0, 0x0, 0xc000b7af60, 0x0, 0x0, ...)
	github.com/caddyserver/caddy/v2@v2.4.0-beta.1/caddyconfig/httpcaddyfile/tlsapp.go:356 +0xed8
github.com/caddyserver/caddy/v2/caddyconfig/httpcaddyfile.ServerType.Setup(0xc0007a1000, 0x43, 0x80, 0xc0006bb800, 0xc0001e5080, 0x5, 0x8, 0x0, 0xc0006bb9b0, 0x0)
	github.com/caddyserver/caddy/v2@v2.4.0-beta.1/caddyconfig/httpcaddyfile/httptype.go:218 +0x1095
github.com/caddyserver/caddy/v2/caddyconfig/caddyfile.Adapter.Adapt(0x1a26a40, 0x2414708, 0xc000374840, 0xb5, 0x2b5, 0xc0006bb800, 0xc0003ba360, 0x62000000016fda40, 0xc000bdf598, 0x626ade64124565c6, ...)
	github.com/caddyserver/caddy/v2@v2.4.0-beta.1/caddyconfig/caddyfile/adapter.go:50 +0x162
github.com/caddyserver/caddy/v2/cmd.loadConfig(0x7ffd96f44ec0, 0x14, 0x7ffd96f44edf, 0x9, 0xc000194330, 0x403e33, 0xc000bdfac0, 0x756ea1a0777002, 0x3026626bc76bb325, 0x40e3b8, ...)
	github.com/caddyserver/caddy/v2@v2.4.0-beta.1/cmd/main.go:177 +0x644
github.com/caddyserver/caddy/v2/cmd.cmdRun(0xc0001843c0, 0x0, 0x0, 0x0)
	github.com/caddyserver/caddy/v2@v2.4.0-beta.1/cmd/commandfuncs.go:206 +0x805
github.com/caddyserver/caddy/v2/cmd.Main()
	github.com/caddyserver/caddy/v2@v2.4.0-beta.1/cmd/main.go:85 +0x248
main.main()
	caddy/main.go:12 +0x25
```

Dockerfile:
```
FROM caddy:2-builder-alpine AS builder

RUN xcaddy build \
    --with github.com/caddy-dns/cloudflare

FROM caddy:2-alpine

COPY --from=builder /usr/bin/caddy /usr/bin/caddy
```

docker-compose.yml:
```
version: '3.7'

services:
  caddy:
    image: caddy:2-dns-cloudflare
    build: .
    container_name: caddy
    restart: always
    environment:
      XDG_DATA_HOME: /shared/data
    ports:
      - '80:80'
      - '443:443'
    volumes:
      - /path/to/my/config/caddy:/etc/caddy
      - config:/config
      - shared:/shared
```

**Comment by francislavoie** [comment]
Thanks for opening an issue! We'll look into this.

It's not immediately clear to us what is going on, so we'll need your help to understand it better.

Ideally, we need to be able to reproduce the bug _in the most minimal way possible_. This allows us to write regression tests to verify the fix is working. If we can't reproduce it, then you'll have to test our changes for us until it's fixed -- and then we can't add test cases, either.

I've attached a template below that will help make this easier and faster! It will ask for some information you've already provided; that's OK, just fill it out the best you can. :+1:

I've also included some helpful tips below the template. Feel free to let me know if you have any questions!

Thank you again for your report, we look forward to resolving it!

## Template

<pre><code>## 1. Environment

### 1a. Operating system and version

```
paste here
```


### 1b. Caddy version (run `caddy version` or paste commit SHA)

```
paste here
```


### 1c. Go version (if building Caddy from source; run `go version`)

```
paste here
```


## 2. Description

### 2a. What happens (briefly explain what is wrong)




### 2b. Why it's a bug (if it's not obvious)




### 2c. Log output

```
paste terminal output or logs here
```



### 2d. Workaround(s)




### 2e. Relevant links




## 3. Tutorial (minimal steps to reproduce the bug)




</code></pre>

## Helpful tips

1. **Environment**: Please fill out your OS and Caddy versions, even if you don't think they are relevant. (They are _always_ relevant.) If you built Caddy from source, provide the commit SHA and specify your exact Go version.

2. **Description:** Describe at a high level what the bug is. What happens? Why is it a bug? Not all bugs are obvious, so convince readers that it's actually a bug.
	- **2c) Log output:** Paste terminal output and/or complete logs in a code block. [DO NOT REDACT INFORMATION except for credentials.](https://caddy.community/t/how-to-get-help-with-caddy-more-effectively/5222)
	- **2d) Workaround:** What are you doing to work around the problem in the meantime? This can help others who encounter the same problem, until we implement a fix.
	- **2e) Relevant links:** Please link to any related issues, pull requests, docs, and/or discussion. This can add crucial context to your report.

3. **Tutorial:** What are the _minimum required specific steps_ someone needs to take in order to experience the same bug? Your goal here is to make sure that anyone else can have the same experience with the bug as you do. You are writing a tutorial, so make sure to carry it out yourself before posting it. Please:
	- Start with an empty config. Add _only_ the lines/parameters that are _absolutely required_ to reproduce the bug.
	- Do not run Caddy inside containers.
	- Run Caddy manually in your terminal; do not use systemd or other init systems.
	- If making HTTP requests, avoid web browsers. Use a simpler HTTP client instead, like `curl`.
	- Do not redact any information from your config (except credentials). Domain names are public knowledge and [often necessary for quick resolution of an issue](https://caddy.community/t/how-to-get-help-with-caddy-more-effectively/5222)!
	- Note that ignoring this advice may result in delays, or even in your issue being closed. 😞 Only actionable issues are kept open, and if there is not enough information or clarity to reproduce the bug, then the report is not actionable.

Example of a tutorial:

<blockquote>Create a config file:

```
{ ... }
```

Open terminal and run Caddy:

```
$ caddy ...
```

Make an HTTP request:

```
$ curl ...
```

Notice that the result is ___ but it should be ___.</blockquote>

**Comment by psrcek** [comment]
## 1. Environment

### 1a. Operating system and version

```
$ uname -a
Linux docker-host-01 4.19.0-16-amd64 #1 SMP Debian 4.19.181-1 (2021-03-19) x86_64 GNU/Linux
$ docker -v
Docker version 20.10.5, build 55c4c88
```


### 1b. Caddy version (run `caddy version` or paste commit SHA)

```
caddy: v2.3.0
xcaddy: v0.1.8
```

### 1c. Go version (if building Caddy from source; run `go version`)

```
1.15.11
```

### 1d. Docker image (where I extracted the above info)

Note: this image is built using docker-compose with the information provided in the OP. If you wish I can upload the image somewhere for you to test with.

<details>
  <summary>Click to expand</summary>

```
$ docker image inspect caddy:2-dns-cloudflare
[
    {
        "Id": "sha256:4a1482aa922d6b11bd0bc89b467fb03ad602053480f6a3c479ce10a2d135cb35",
        "RepoTags": [
            "caddy:2-dns-cloudflare"
        ],
        "RepoDigests": [],
        "Parent": "sha256:a399f1211b45b8dbf10005ab3deec678552c502ac66a63ea3027a2ac30051414",
        "Comment": "",
        "Created": "2021-04-02T11:23:25.497704234Z",
        "Container": "",
        "ContainerConfig": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "ExposedPorts": {
                "2019/tcp": {},
                "443/tcp": {},
                "80/tcp": {}
            },
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "CADDY_VERSION=v2.3.0",
                "XDG_CONFIG_HOME=/config",
                "XDG_DATA_HOME=/data"
            ],
            "Cmd": [
                "/bin/sh",
                "-c",
                "#(nop) COPY file:dda93cef83c7a88198f6b6d40e4e8f5842bd328bb57f910f45099dc9c5c8efa0 in /usr/bin/caddy "
            ],
            "Image": "sha256:a399f1211b45b8dbf10005ab3deec678552c502ac66a63ea3027a2ac30051414",
            "Volumes": {
                "/config": {},
                "/data": {}
            },
            "WorkingDir": "/srv",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": {
                "org.opencontainers.image.description": "a powerful, enterprise-ready, open source web server with automatic HTTPS written in Go",
                "org.opencontainers.image.documentation": "https://caddyserver.com/docs",
                "org.opencontainers.image.licenses": "Apache-2.0",
                "org.opencontainers.image.source": "https://github.com/caddyserver/caddy-docker",
                "org.opencontainers.image.title": "Caddy",
                "org.opencontainers.image.url": "https://caddyserver.com",
                "org.opencontainers.image.vendor": "Light Code Labs",
                "org.opencontainers.image.version": "v2.3.0"
            }
        },
        "DockerVersion": "20.10.5",
        "Author": "",
        "Config": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "ExposedPorts": {
                "2019/tcp": {},
                "443/tcp": {},
                "80/tcp": {}
            },
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "CADDY_VERSION=v2.3.0",
                "XDG_CONFIG_HOME=/config",
                "XDG_DATA_HOME=/data"
            ],
            "Cmd": [
                "caddy",
                "run",
                "--config",
                "/etc/caddy/Caddyfile",
                "--adapter",
                "caddyfile"
            ],
            "Image": "sha256:a399f1211b45b8dbf10005ab3deec678552c502ac66a63ea3027a2ac30051414",
            "Volumes": {
                "/config": {},
                "/data": {}
            },
            "WorkingDir": "/srv",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": {
                "org.opencontainers.image.description": "a powerful, enterprise-ready, open source web server with automatic HTTPS written in Go",
                "org.opencontainers.image.documentation": "https://caddyserver.com/docs",
                "org.opencontainers.image.licenses": "Apache-2.0",
                "org.opencontainers.image.source": "https://github.com/caddyserver/caddy-docker",
                "org.opencontainers.image.title": "Caddy",
                "org.opencontainers.image.url": "https://caddyserver.com",
                "org.opencontainers.image.vendor": "Light Code Labs",
                "org.opencontainers.image.version": "v2.3.0"
            }
        },
        "Architecture": "amd64",
        "Os": "linux",
        "Size": 72889918,
        "VirtualSize": 72889918,
        "GraphDriver": {
            "Data": {
                "LowerDir": "/var/lib/docker/overlay2/a2a864eedfa34093de5c3c5cad61a1f106e3fced22e47ef7e71d76c0b87f7703/diff:/var/lib/docker/overlay2/accb12be927d71666264222188ca1266569255f66d1f3035bbe24a193c48ec4a/diff:/var/lib/docker/overlay2/468671df16def7d552fafb69baabcdcdb03ecc5de488ee9f8400b59c677b4c3d/diff:/var/lib/docker/overlay2/0ecd833c2e0a410b8e852eae0719223786a7f7054c6ad4fd9ef27562575abc47/diff:/var/lib/docker/overlay2/319ebcd7d9848aa49894a4dd537142cbfafcb501aed455a94bae550dc209263b/diff",
                "MergedDir": "/var/lib/docker/overlay2/0ad363d04a799d151b7f5fe30354729ca9e66f6797a07dd4c4e2d5ae05d44c7a/merged",
                "UpperDir": "/var/lib/docker/overlay2/0ad363d04a799d151b7f5fe30354729ca9e66f6797a07dd4c4e2d5ae05d44c7a/diff",
                "WorkDir": "/var/lib/docker/overlay2/0ad363d04a799d151b7f5fe30354729ca9e66f6797a07dd4c4e2d5ae05d44c7a/work"
            },
            "Name": "overlay2"
        },
        "RootFS": {
            "Type": "layers",
            "Layers": [
                "sha256:fa29914ad24cae109d0e1ba325e273c94b2e3de471c2b87ba7feddf245318c3e",
                "sha256:3177f360d8a69473b20a7956aa798fd7955126bfd8680b069ac348c93e6def6f",
                "sha256:6c5e304de64bb6328a88228fda955ae40a6f0af949926246cb46ea2f783ffe4b",
                "sha256:e424f7d8963e4412353c724936f9cbc77c2b7dc89a2ed53d04aaf80d86a61330",
                "sha256:33055f1567833c01fbdcab1ee56ad447d41d8a95fefd3a24f4255cca60375a76",
                "sha256:cfa8b57f7f9da95c47f4ea35776173168c9439d6c43ad284c4fedbf4e2943930"
            ]
        },
        "Metadata": {
            "LastTagTime": "2021-04-02T13:23:28.839944024+02:00"
        }
    }
]
```
</details>


## 2. Description

### 2a. What happens (briefly explain what is wrong)

Caddy will not start, error is printed to stdout.

### 2c. Log output

```
{"level":"info","ts":1617283700.1680074,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":"caddyfile"}
panic: runtime error: index out of range [-1]

goroutine 1 [running]:
github.com/caddyserver/caddy/v2/caddyconfig/httpcaddyfile.consolidateAutomationPolicies(0xc00052e800, 0x3, 0x40, 0x0, 0x0, 0xc0006e0de0)
	github.com/caddyserver/caddy/v2@v2.4.0-beta.1/caddyconfig/httpcaddyfile/tlsapp.go:498 +0xc4a
github.com/caddyserver/caddy/v2/caddyconfig/httpcaddyfile.ServerType.buildTLSApp(0xc0001f25a0, 0x5, 0x5, 0xc0006bb800, 0x0, 0x0, 0x0, 0xc000b7af60, 0x0, 0x0, ...)
	github.com/caddyserver/caddy/v2@v2.4.0-beta.1/caddyconfig/httpcaddyfile/tlsapp.go:356 +0xed8
github.com/caddyserver/caddy/v2/caddyconfig/httpcaddyfile.ServerType.Setup(0xc0007a1000, 0x43, 0x80, 0xc0006bb800, 0xc0001e5080, 0x5, 0x8, 0x0, 0xc0006bb9b0, 0x0)
	github.com/caddyserver/caddy/v2@v2.4.0-beta.1/caddyconfig/httpcaddyfile/httptype.go:218 +0x1095
github.com/caddyserver/caddy/v2/caddyconfig/caddyfile.Adapter.Adapt(0x1a26a40, 0x2414708, 0xc000374840, 0xb5, 0x2b5, 0xc0006bb800, 0xc0003ba360, 0x62000000016fda40, 0xc000bdf598, 0x626ade64124565c6, ...)
	github.com/caddyserver/caddy/v2@v2.4.0-beta.1/caddyconfig/caddyfile/adapter.go:50 +0x162
github.com/caddyserver/caddy/v2/cmd.loadConfig(0x7ffd96f44ec0, 0x14, 0x7ffd96f44edf, 0x9, 0xc000194330, 0x403e33, 0xc000bdfac0, 0x756ea1a0777002, 0x3026626bc76bb325, 0x40e3b8, ...)
	github.com/caddyserver/caddy/v2@v2.4.0-beta.1/cmd/main.go:177 +0x644
github.com/caddyserver/caddy/v2/cmd.cmdRun(0xc0001843c0, 0x0, 0x0, 0x0)
	github.com/caddyserver/caddy/v2@v2.4.0-beta.1/cmd/commandfuncs.go:206 +0x805
github.com/caddyserver/caddy/v2/cmd.Main()
	github.com/caddyserver/caddy/v2@v2.4.0-beta.1/cmd/main.go:85 +0x248
main.main()
	caddy/main.go:12 +0x25
```

### 2d. Workaround(s)

I don't have any


### 2e. Relevant links

Have not found any.


## 3. Tutorial (minimal steps to reproduce the bug)

1. Use the following config:

```Caddyfile
{
	email my.email@example.com
}

:82 {
	redir https://example.com{uri}
}

:83 {
	redir https://example.com{uri}
}

:84 {
	redir https://example.com{uri}
}

abc.de {
	redir https://example.com{uri}
}
```

2. start caddy.
3. observe above error

## 4. Note

I would like to add that the above Caddyfile is the **MINIMUM** I was able to use to reproduce this error.

- Removing any block (including global) in the Caddyfile makes it start, when I put it back it's broken again.
- Changing abc.de to http://abc.de makes it start. (yes I made that up, but it happens with my actual domain aswell. Both domains match the regex `/[a-z]{3}\.[a-z]{2}/`)
- Changing the contents of the blocks (to e.g. a reverse_proxy) doesn't do anything.

**Comment by francislavoie** [comment]
Okay, a few things to comment on:

- You say you're using `v2.3.0`, but actually `xcaddy` built with `v2.4.0-beta.1` as you can see in the panic log. This is due to a known issue with how Go dependency edits end up stepping on eachother; see https://github.com/caddyserver/xcaddy/issues/54
- If you try to do a fresh build, you should now get `v2.4.0-beta.2` instead, it was just released last night. It resolves some issues in compatibility between caddy and the cloudflare plugin that existed in the `v2.4.0-beta.1` version.
- Why are you doing this in your `docker-compose.yml`? The volumes by default are `/data` and `/config`, I see no reason to override that. See the docs https://hub.docker.com/_/caddy

      environment:
        XDG_DATA_HOME: /shared/data
      volumes:
        - /path/to/my/config/caddy:/etc/caddy
        - config:/config
        - shared:/shared

All that said, I can reproduce the panic with that Caddyfile with the latest commit on master. I'll dig into it.

**Comment by francislavoie** [comment]
I proposed a PR to fix it, #4104

In the meantime, you could do something like this instead, which doesn't produce the panic:

```
:82, :83, :84 {
	redir https://example.com{uri}
}

abc.de {
	tls my.email@example.com
	redir https://example.com{uri}
}
```

**Comment by psrcek** [comment]
> In the meantime, you could do something like this instead, which doesn't produce the panic

No I can't because there's different directives on the ports, I've just reduced it to the bare minimum for this issue.

Also, thanks for fixing this so fast! This is exactly why I love using caddy! :heart:

**Comment by psrcek** [comment]
Is there any easy way for me to run a build with this fix since this is currently breaking our production config? How soon is this going to be released into a beta? I'm assuming it isn't included in `v2.4.0-beta.2` since I've tried that and it's still broken.

**Comment by francislavoie** [comment]
You can use xcaddy to build from source https://caddyserver.com/docs/build just specify the commit hash or "master" as the first argument to `xcaddy build`
